src/components/EggheadPlayer/use-egghead-player.ts

```diff
@@ -0,0 +1,213 @@
+
+  if (!canTrack) {
+    return
+  }
+
+  lessonProgress[lesson.slug] = roundedProgress
+
+  if (lesson.lesson_view_url) {
+    if (roundedProgress === 30) {
+      track('started lesson', {
+        lesson: lesson.slug,
+        tags: lesson.tags.map((tag) => tag.slug),
+      })
+    }
+
+    storeProgress(roundedProgress, lesson)
+  } else {
+    const segments = lessonProgress[lesson.slug] / 30
+    let lessonViews = cookies.get('egghead-lessonViews') || {}
+    lessonViews[lesson.id] = segments
+
+    cookies.set('egghead-lessonViews', lessonViews)
+  }
+}
+
+let emailCaptureCookie: any = {
+  watchCount: 0,
+}
+
+const setEmailCaptureCookie = (captureCookie = {}) => {
+  cookies.set('egghead-email', captureCookie)
+  emailCaptureCookie = cookies.get('egghead-email')
+}
+
+const getOrCreateLessonView = (lesson, collection?) => {
+  return axios
+    .post(
+      lesson.lesson_view_url,
+      toLessonViewParams({lesson, collection}),
+      getOptions(),
+    )
+    .then(({data}) => data)
+}
+
+const trackPercentComplete = (lessonView) => {
+  const series = lessonView.series
+  var percents = [0.1, 0.25, 0.5, 0.75]
+  if (series && series.progress && series.published_lesson_count) {
+    percents.forEach((percent) => {
+      var trackPercent =
+          Math.ceil(series.published_lesson_count * percent) ===
+          series.progress.completed_lesson_count,
+        percentCompleted = Math.floor(percent * 100)
+      if (trackPercent) {
+        track('progress in course', {
+          course: series.slug,
+          percent_completed: percentCompleted,
+        })
+      }
+    })
+  }
+  return lessonView
+}
+
+const trackStartingCourse = (lessonView) => {
+  const series = lessonView.series
+  const progress = series && lessonView.series.progress
+
+  if (progress && progress.completed_lesson_count === 0) {
+    track('started course', {
+      first_lesson: lessonView.lesson_slug,
+      lesson_count: series.published_lesson_count,
+      course: lessonView.series.slug,
+    })
+  }
+  return lessonView
+}
+
+const completeLesson = (lessonView) =>
+  axios.put(lessonView.complete_url, {}, getOptions()).then(({data}) => {
+    const lessonView = data
+    const series = lessonView.series
+    const courseProgress = series && lessonView.series.progress
+
+    if (courseProgress && courseProgress.is_completed) {
+      track('finished course', {
+        final_lesson: lessonView.lesson_slug,
+        course: lessonView.series.slug,
+        enhanced_transcripts_complete:
+          series.lesson_count - series.enhanced_transcript_count === 0,
+        lesson_count: series.published_lesson_count,
+      })
+    }
+
+    return lessonView
+  })
+
+const onComplete = (lesson, collection?) => {
+  return getOrCreateLessonView(lesson, collection)
+    .then(trackStartingCourse)
+    .then(completeLesson)
+    .then(trackPercentComplete)
+}
+
+const onEnded = (lesson) => async () => {
```

Comment:
this was almost a full port. in the egghead-rails version, it was managing some view state as well which is omitted here.